adapters/richaudience/richaudience.go

```
@@ -0,0 +1,349 @@
+	Device richaudienceDevice `json:"device,omitempty"`
+	Site   richaudienceSite   `json:"site,omitempty"`
+	Test   int8               `json:"test,omitempty"`
+}
+
+type richaudienceUser struct {
+	BuyerUID string              `json:"buyeruid,omitempty"`
+	Ext      richaudienceUserExt `json:"ext,omitempty"`
+}
+
+type richaudienceUserExt struct {
+	Eids []openrtb_ext.ExtUserEid `json:"eids,omitempty"`
+	GDPR string                   `json:"gdpr,omitempty"`
+}
+
+type richaudienceDevice struct {
+	//GEO only Java, IFA
+	IP  string `json:"ip,omitempty"`
+	Lmt int8   `json:"lmt,omitempty"`
+	DNT int8   `json:"dnt,omitempty"`
+	UA  string `json:"ua,omitempty"`
+}
+type richaudienceSite struct {
+	//Site: Cat IAB
+	Domain string `json:"domain,omitempty"`
+	Page   string `json:"page,omitempty"`
+}
+
+// Builder builds a new instance of the Foo adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+
+	template, err := template.New("endpointTemplate").Parse(config.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse endpoint url template: %v", err)
+	}
+
+	if config.Endpoint != "" && strings.Contains(config.Endpoint, "richaudience") {
+		bidder := &RichaudienceAdapter{
+			EndpointTemplate: *template,
+		}
+
+		return bidder, nil
+	}
+
+	return nil, nil
+}
+
+func (a *RichaudienceAdapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	richaudienceRequest := richaudienceRequest{}
+
+	raiHeaders := http.Header{}
+	//Global Vars
+	secure := int8(0)
+	//Object IMP
+	resImps := make([]openrtb2.Imp, 0, len(request.Imp))
+	raiErrors := make([]error, 0, len(request.Imp))
+
+	var regsExt *openrtb_ext.ExtRegs
+	if request.Regs != nil && request.Regs.Ext != nil {
+		if err := json.Unmarshal(request.Regs.Ext, &regsExt); err != nil {
+			fmt.Println(err)
```

Please return error rather than print it. Please remove all Print statements.

---

in next commit i have solved:

i have all remove printf errors, please check it